linera-service/src/grpc_proxy.rs

```diff
@@ -20,15 +21,77 @@ use linera_rpc::{
     },
     grpc_pool::ConnectionPool,
 };
-use std::{fmt::Debug, net::SocketAddr, sync::Arc, time::Duration};
+use once_cell::sync::Lazy;
+use prometheus::{register_histogram_vec, HistogramVec};
+use std::{
+    fmt::Debug,
+    net::SocketAddr,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
 use tokio::select;
 use tokio_stream::wrappers::UnboundedReceiverStream;
 use tonic::{
-    transport::{Channel, Server},
+    transport::{Body, Channel, Server},
     Request, Response, Status,
 };
+use tower::{builder::ServiceBuilder, Layer, Service};
 use tracing::{debug, info, instrument};
 
+pub static PROXY_REQUEST_LATENCY: Lazy<HistogramVec> = Lazy::new(|| {
+    register_histogram_vec!(
+        "proxy_request_latency",
+        "Proxy request latency",
+        // Can add labels here
+        &[]
+    )
+    .expect("Counter can be created")
+});
+
+#[derive(Clone)]
+pub struct PrometheusMetricsMiddlewareLayer;
+
+#[derive(Clone)]
+pub struct PrometheusMetricsMiddlewareService<T> {
+    service: T,
+}
+
+impl<S> Layer<S> for PrometheusMetricsMiddlewareLayer {
+    type Service = PrometheusMetricsMiddlewareService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        PrometheusMetricsMiddlewareService { service }
+    }
+}
+
+impl<S> Service<tonic::codegen::http::Request<Body>> for PrometheusMetricsMiddlewareService<S>
+where
+    S::Future: Send + 'static,
+    S: Service<tonic::codegen::http::Request<Body>> + std::marker::Send,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<S::Response, S::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: http::Request<Body>) -> Self::Future {
+        let start = std::time::Instant::now();
+        let future = self.service.call(request);
+        async move {
+            let response = future.await?;
+            PROXY_REQUEST_LATENCY
+                .with_label_values(&[])
+                .observe(start.elapsed().as_secs_f64());
+            Ok(response)
+        }
+        .boxed()
+    }
+}
```

Comment:
Would it be easier to use [`service_fn`](https://docs.rs/tower/latest/tower/fn.service_fn.html) instead of explicitly implementing the trait?

Reply:
Talked with Andreas on Slack, we'll look into this as a follow up